### DISCUSSION TITLE
Forward search with `tarski`

### DISCUSSION THREAD

**Discussion by miquelramirez** [aig-upf/tarski#101] [General]
Following up on the query by Christian Muise on the Planning Community Slack channel, I would like to direct folks to the simple blind search engine for STRIPS planning available in the module [`tarski.search`](https://github.com/aig-upf/tarski/tree/devel/src/tarski/search). 

We *strongly* recommend writing your own search engine in a language more performant than Python.

**Comment by gfrances** [comment]
I haven't been following the channel lately, but it's not even a matter of Python (which of course it also is), it's also that Tarski aims at having a clear design for *representing* planning problems, not so much for solving them. I have added that module `tarski.search` mostly to perform quick explorations of small instances, or to test properties of planning policies, but it's definitely **not a performant solution**, and there are several sources of inefficiency. This is not to say that these inefficiencies cannot be dealt with - actually they can, but we're not there yet. Actually, and I guess this is related to your point, Miquel, ultimately I haven't put too much effort on that because for a performant solution I think it'd be much better to implement that in CPython and simply use Tarski as a Python wrapper.

But: other options for running more efficient searches on a Tarski problem include interfacing with [Pyperplan](https://github.com/aibasel/pyperplan) (I've done that in some projects, and have a few lines of code to easy transform any Tarski state into a Pyperplan state), or, why not, to interface directly with Fast Downward. The advantages of the second are obvious; the disadvantage is that it's a monolithic solution, and you can only ask for a plan, but cannot ask the library useful things such as "give me all nodes at distance at most 4 from this state", etc. If there is interest, I have code for that too as well, which could be easily integrated into Tarski (actually it's a very silly code: just print instance and domain to two temporary files, invoke Fast Downward as a subprocess, parse the output).

**Reply by miquelramirez** [reply]
>  ultimately I haven't put too much effort on that because for a performant solution I think it'd be much better to implement that in CPython and simply use Tarski as a Python wrapper.

Exactly - but rather than `CPython`, what I have found to work **really well** with Modern C++ is [pybind11](https://github.com/pybind/pybind11).

**Comment by haz** [comment]
So our use case here (1) doesn't need the performance; and (2) would suffer entirely the more you tie into C++ and bind. I actually strongly prefer to have a version of tarski that's as close to pure as possible.

We want to embed tarski in the browser, so that we have functional access to all the niceties it offers for things like plugin development. We need to parse PDDL, ground it (working on seeing if Gringo will fly, but more naive grounding that just filters on types would be useful as a backup), find the actions applicable in a state, and run progression.

Need a full plan? Call solver.planning.domains (it's about to be equipped with far more than just lapkt).

So I assume [[this](https://github.com/aig-upf/tarski/blob/devel/src/tarski/search/applicability.py)] is the extent of what we have access to as far as successor generation? Looks like it violates the `s' = (s \ DEL(a)) U ADD(a)` sequence of things. Also, should we be going off of the `devel` branch rather than `master`?

**Reply by miquelramirez** [reply]
> I actually strongly prefer to have a version of tarski that's as close to pure as possible.

`tarski` is and will remain pure Python. Planners that use `tarski` as a front-end not need to be constrained in that way.

> ground it (working on seeing if Gringo will fly, but more naive grounding that just filters on types would be useful as a backup), 

There's a fallback already in place, the `tarski.grounding.naive` module and some code examples.

> So I assume [this] is the extent of what we have access to as far as successor generation? Looks like it violates the s' = (s \ DEL(a)) U ADD(a) sequence of things. Also, should we be going off of the devel branch rather than master?

That was a stealth bug report 👍

**Comment by gfrances** [comment]
Ouuch great catch, Christian! That looks like an unused piece of code that shouldn't be there, actually what I use is [this other method](https://github.com/aig-upf/tarski/blob/devel/src/tarski/search/model.py#L23) that does indeed respect the add-after-delete semantics.
But definitely the other method only adds confusion (looking at the Git log, very likely I reimplemented the same thing with a few months' difference :-))

I'd say we can fix the `resource vs. pypsutil` issue discussed on the chat, then we can upload a new version to pypi
and you can go against `master`. What do you think?
Of course you're free to go against `devel`, but sometimes we break things there.

G.

**Reply by haz** [reply]
Well my concern is with functionality. In particular, progression be gone?
https://github.com/aig-upf/tarski/blob/master/src/tarski/search/model.py#L23

**Reply by miquelramirez** [reply]
The design of the `search` module seems to me to be a work in progress. @gfrances was pointing you to a class that is unrolling the state-transition model forward. The change is that there is no longer an abstract class `ForwardSearchModel`. It has been replaced with a `GroundForwardSearchModel` one, probably as @gfrances can't see any time soon a `LiftedForwardSearchModel` happening.

**Reply by haz** [reply]
Does that really have to be so? I could see it being _extremely_ useful to have lifted successor generators and progression -- not for planning purposes, but for avoiding the full problem grounding. Either because the problem is huge or gringo just doesn't fly browser side.

**Reply by miquelramirez** [reply]
I don't know what was the rationale for the change @haz. If you have a pointer for a lifted successor generation algorithm that scales up and leads to solve a truly *significant* number of instances in addition to what blind search on a grounded representation can do, I am all ears.

**Reply by gfrances** [reply]
What do you mean, Miquel, are you not following my papers?!? 

https://ai.dmi.unibas.ch/papers/correa-et-al-icaps2020.pdf

Ok ok, I hear you, depends on how you define "truly significant" :-))

On a more serious note, I'm not sure exactly what we're discussing here.
The `devel` branch has an abstract `SearchModel` class (could perhaps better be named `ForwardSearchModel`, but that's another story). Adding a lifted search model should be relatively easy; adding another abstract base class to distinguish between search, forward search, and backward search, should be doable too. In summary, everything is pretty open, but we don't have anything else at the moment beyond a simple, non-optimized ground forward search.

It'd be good to add some tests to that too, I'll try to do that later today.

**Reply by miquelramirez** [reply]
Hi Guillem,

The discussion followed from observing that there used to be hierarchy of classes, which is the same you use to have in `fs` iirc. In `devel` we have a less ambitious design, which I guess is a work in progress.

**Reply by haz** [reply]
Well originally my thought was on non-search related progression -- e.g., interactive state-space exploration / visualization. I'm less all-in on it now, since our core visualization for each node is the hmax / hadd computation, and we need all the ground actions/fluents for that anyways :P

**Reply by miquelramirez** [reply]
> ., interactive state-space exploration / visualization

That makes perfect sense - offering a tool for the person developing the specification to verify if a given sequence of actions produces a particular property looks really useful.

**Reply by miquelramirez** [reply]
But that kind of overlaps with Planimation perhaps?

**Reply by haz** [reply]
Only partially. Think more about the search space visualization we typically see -- nodes are states, edges are actions. So not visualizing the domain specifically.

From that view, we double-click a node to bring up a custom visualization for the heuristic computation.

All this kind of works, but will be far more robust and scalable if we can run in on top of tarski grounding things.

**Reply by miquelramirez** [reply]
Something along the lines of the visualization plug-in for `gecode`?

https://research.monash.edu/files/257776846/3566886_oa.pdf

**Reply by meneguzzi** [reply]
Sort of barging in here, but given the context, I reckon he means something along the lines of Mau's search space visualizer in Web Planner: https://link.springer.com/chapter/10.1007/978-3-030-38561-3_11 

I can see how expanding things on demand as you render rather than generating a huge file expanded on one go could be useful.

**Comment by miquelramirez** [comment]
Not barging in at all @Felipe Meneguzzi ***@***.***> that
is a useful pointer, and indeed, the planning analogue of what I was
thinking of.

**Reply by haz** [reply]
Yep, you guys are spot on. Here's the discussion between Mau and I on it: https://github.com/AI-Planning/heuristic-viz/issues/6

**Reply by meneguzzi** [reply]
Need access!

**Reply by meneguzzi** [reply]
And Miquel, indeed, I noticed the similarities, interesting paper indeed.

**Reply by haz** [reply]
How are you not in the `ai-planning` org already? Oy...

**Reply by meneguzzi** [reply]
I feel you don't love me anymore ;-)

**Reply by haz** [reply]
Bah. I put the call out way back when -- you just didn't answer!

Hope you can see it now.
